membership-attribute-service/app/controllers/AccountController.scala

```diff
@@ -269,23 +303,72 @@ class AccountController(commonActions: CommonActions, override val controllerCom
       }
     } yield filteredIfApplicable
 
+  def getGiftPaymentDetails(giftSub: Subscription[SubscriptionPlan.Paid]) = PaymentDetails(
+    pendingCancellation = giftSub.isCancelled,
+    chargedThroughDate = None,
+    startDate = giftSub.startDate,
+    customerAcceptanceDate = giftSub.startDate,
+    nextPaymentPrice = None,
+    lastPaymentDate = None,
+    nextPaymentDate = None,
+    termEndDate = giftSub.termEndDate,
+    pendingAmendment = false,
+    paymentMethod = None,
+    plan = PersonalPlan(
+      name = giftSub.plan.productName,
+      price = Price(0f, giftSub.plan.charges.currencies.head),
+      interval = BillingPeriod.Year.noun
+    ),
+    subscriberId = giftSub.name.get,
+    remainingTrialLength = 0
+  )
+
+  def getPaymentDetailsForSub(
+    maybeUserId: Option[String],
+    freeOrPaidSub: Subscription[SubscriptionPlan.Free] \/ Subscription[SubscriptionPlan.Paid],
+    paymentService: PaymentService
+  ): Future[PaymentDetails] = freeOrPaidSub match {
+      case \/-(paidSub) if paidSub.gifteeIdentityId != maybeUserId =>
+        paymentService.paymentDetails(freeOrPaidSub, defaultMandateIdIfApplicable = Some(""))
```

Comment:
so the original functionality is only called if the giftee identity id is not the same as the user logged in.
I think this will work, but it is a bit obscured given that this is protecting someone else's payment information.

Ideally I'd prefer to make the decision what kind of payment info to insert before combining the giftee subscription with the self-owned subscriptions.  So move this decision into the `allCurrentSubscriptions`.  Then there is no mixing of the code paths for getting payment details and getting giftee subscriptions.

Failing that we can make the pattern match check more explicitly that it's a gift ReaderType and also add a test to make sure the overall logic doesn't get broken.

Reply:
Ok I've swapped things around a bit to make it more obvious, I've also added a check on the ReaderType and some tests.